File: jps/jps-builders/src/org/jetbrains/jps/incremental/storage/Xxh3HashingService.java
// Copyright 2000-2023 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.jps.incremental.storage;

import com.intellij.util.lang.Xxh3;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;

final class Xxh3HashingService {
  static long getStringHash(@NotNull String hashableString) {
    return Xxh3.hash(hashableString);
  }

  static long getFileHash(@NotNull File file) throws IOException {
    long fileHash;
    try (FileInputStream fis = new FileInputStream(file)) {
      fileHash = Xxh3.hash(fis, (int)file.length());
    }
    return fileHash;
  }

  static long hashLongs(long... values) {
    return Xxh3.hashLongs(values);
  }
}

File: jps/jps-builders/src/org/jetbrains/jps/incremental/storage/Xxh3HashingService.kt
// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.jps.incremental.storage

import com.intellij.util.lang.Xxh3
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path

@Throws(IOException::class)
fun getFileHash(file: File): Long {
  var fileHash: Long
  FileInputStream(file).use { fis ->
    fileHash = Xxh3.hash(fis, Math.toIntExact(file.length()))
  }
  return fileHash
}

@Throws(IOException::class)
fun getFileHash(file: Path): Long {
  var fileHash: Long
  Files.newInputStream(file).use { fis ->
    fileHash = Xxh3.hash(fis, Math.toIntExact(Files.size(file)))
  }
  return fileHash
}
